diff --git a/sim/simx/core.h b/sim/simx/core.h
index a8b674d..f6be86f 100644
--- a/sim/simx/core.h
+++ b/sim/simx/core.h
@@ -20,6 +20,7 @@
 #include "pipeline.h"
 #include "cache_sim.h"
 #include "local_mem.h"
+#include "tensor_mem.h"
 #include "ibuffer.h"
 #include "scoreboard.h"
 
@@ -147,6 +148,10 @@ public:
     return local_mem_;
   }
 
+  const TensorMem::Ptr& tensor_mem() const {
+    return tensor_mem_;
+  }
+
   const MemCoalescer::Ptr& mem_coalescer(uint32_t idx) const {
     return mem_coalescers_.at(idx);
   }
@@ -208,6 +213,7 @@ private:
   std::vector<Dispatcher::Ptr> dispatchers_;
   std::vector<FuncUnit::Ptr> func_units_;
   LocalMem::Ptr local_mem_;
+  TensorMem::Ptr tensor_mem_;
   std::vector<LocalMemSwitch::Ptr> lmem_switch_;
   std::vector<MemCoalescer::Ptr> mem_coalescers_;
 
diff --git a/sim/simx/emulator.cpp b/sim/simx/emulator.cpp
index 3eb62f9..658a750 100644
--- a/sim/simx/emulator.cpp
+++ b/sim/simx/emulator.cpp
@@ -313,6 +313,8 @@ void Emulator::dcache_read(void *data, uint64_t addr, uint32_t size) {
   auto type = get_addr_type(addr);
   if (type == AddrType::Shared) {
     core_->local_mem()->read(data, addr, size);
+  } else if (type == AddrType::Tensor) {
+    core_->tensor_mem()->read(data, addr, size);
   } else {
     try
     {
@@ -331,6 +333,8 @@ void Emulator::dcache_read(void *data, uint64_t addr, uint32_t size) {
   auto type = get_addr_type(addr);
   if (type == AddrType::Shared) {
     core_->local_mem()->read(data, addr, size);
+  } else if (type == AddrType::Tensor) {
+    core_->tensor_mem()->read(data, addr, size);
   } else {
     mmu_.read(data, addr, size, 0);
   }
@@ -348,6 +352,8 @@ void Emulator::dcache_write(const void* data, uint64_t addr, uint32_t size) {
   } else {
     if (type == AddrType::Shared) {
       core_->local_mem()->write(data, addr, size);
+    } else if (type == AddrType::Tensor) {
+      core_->tensor_mem()->write(data, addr, size);
     } else {
       try
       {
@@ -372,6 +378,8 @@ void Emulator::dcache_write(const void* data, uint64_t addr, uint32_t size) {
   } else {
     if (type == AddrType::Shared) {
       core_->local_mem()->write(data, addr, size);
+    } else if (type == AddrType::Tensor) {
+      core_->tensor_mem()->write(data, addr, size);
     } else {
       mmu_.write(data, addr, size, 0);
     }
diff --git a/sim/simx/tensor_mem.cpp b/sim/simx/tensor_mem.cpp
new file mode 100644
index 0000000..c1f2b2e
--- /dev/null
+++ b/sim/simx/tensor_mem.cpp
@@ -0,0 +1,127 @@
+#include "tensor_mem.h"
+#include "core.h"
+#include <bitmanip.h>
+#include <vector>
+#include "types.h"
+
+using namespace vortex;
+
+class TensorMem::Impl {
+protected:
+	TensorMem* simobject_;
+	Config    config_;
+	RAM       ram_; // TODO: is RAM the correct MemDevice here?
+	uint32_t 	line_bits_;
+	MemCrossBar::Ptr mem_xbar_;
+	mutable PerfStats perf_stats_;
+
+	uint64_t to_local_addr(uint64_t addr) {
+		return bit_getw(addr, 0, line_bits_-1);
+	}
+
+public:
+	Impl(TensorMem* simobject, const Config& config)
+		: simobject_(simobject)
+		, config_(config)
+		, ram_(config.capacity)
+	{
+		uint32_t total_lines = config.capacity / config.line_size;
+		line_bits_ = log2ceil(total_lines);
+
+		char sname[100];
+		snprintf(sname, 100, "%s-xbar", simobject->name().c_str());
+		uint32_t lg2_line_size = log2ceil(config_.line_size);
+		uint32_t num_banks = 1 << config.B;
+		mem_xbar_ = MemCrossBar::Create(sname, ArbiterType::Priority, config.num_reqs, num_banks,
+		 [lg2_line_size, num_banks](const MemCrossBar::ReqType& req) {
+    	// Custom logic to calculate the output index using bank interleaving
+			return (uint32_t)((req.addr >> lg2_line_size) & (num_banks-1));
+		});
+		for (uint32_t i = 0; i < config.num_reqs; ++i) {
+			simobject->Inputs.at(i).bind(&mem_xbar_->ReqIn.at(i));
+			mem_xbar_->RspIn.at(i).bind(&simobject->Outputs.at(i));
+		}
+	}
+
+	virtual ~Impl() {}
+
+	void reset() {
+		perf_stats_ = PerfStats();
+	}
+
+	void read(void* data, uint64_t addr, uint32_t size) {
+		auto l_addr = to_local_addr(addr);
+		DPH(3, "Local Mem addr=0x" << std::hex << l_addr << std::dec << std::endl);
+		ram_.read(data, l_addr, size);
+	}
+
+	void write(const void* data, uint64_t addr, uint32_t size) {
+		auto l_addr = to_local_addr(addr);
+		DPH(3, "Local Mem addr=0x" << std::hex << l_addr << std::dec << std::endl);
+		ram_.write(data, l_addr, size);
+	}
+
+	void tick() {
+		// process bank requets from xbar
+		uint32_t num_banks = (1 << config_.B);
+		for (uint32_t i = 0; i < num_banks; ++i) {
+			auto& xbar_req_out = mem_xbar_->ReqOut.at(i);
+			if (xbar_req_out.empty())
+				continue;
+
+			auto& bank_req = xbar_req_out.front();
+			DT(4, simobject_->name() << "-bank" << i << "-req : " << bank_req);
+
+			if (!bank_req.write || config_.write_reponse) {
+				// send xbar response
+				MemRsp bank_rsp{bank_req.tag, bank_req.cid, bank_req.uuid};
+				mem_xbar_->RspOut.at(i).push(bank_rsp);
+			}
+
+			// update perf counters
+			perf_stats_.reads += !bank_req.write;
+			perf_stats_.writes += bank_req.write;
+
+			// remove input
+			xbar_req_out.pop();
+		}
+	}
+
+	const PerfStats& perf_stats() const {
+		perf_stats_.bank_stalls = mem_xbar_->collisions();
+		return perf_stats_;
+	}
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+TensorMem::TensorMem(const SimContext& ctx, const char* name, const Config& config)
+	: SimObject<TensorMem>(ctx, name)
+	, Inputs(config.num_reqs, this)
+	, Outputs(config.num_reqs, this)
+	, impl_(new Impl(this, config))
+{}
+
+TensorMem::~TensorMem() {
+  delete impl_;
+}
+
+void TensorMem::reset() {
+  impl_->reset();
+}
+
+void TensorMem::read(void* data, uint64_t addr, uint32_t size) {
+  impl_->read(data, addr, size);
+}
+
+void TensorMem::write(const void* data, uint64_t addr, uint32_t size) {
+  impl_->write(data, addr, size);
+}
+
+void TensorMem::tick() {
+  impl_->tick();
+}
+
+const TensorMem::PerfStats& TensorMem::perf_stats() const {
+  return impl_->perf_stats();
+}
\ No newline at end of file
diff --git a/sim/simx/tensor_mem.h b/sim/simx/tensor_mem.h
new file mode 100644
index 0000000..c4cda78
--- /dev/null
+++ b/sim/simx/tensor_mem.h
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <simobject.h>
+#include "types.h"
+
+namespace vortex {
+
+class TensorMem : public SimObject<TensorMem> {
+public:
+  struct Config {
+    uint32_t capacity;
+    uint32_t line_size;
+    uint32_t num_reqs;
+    uint32_t B; // log2 number of banks
+    bool write_reponse;
+  };
+
+  struct PerfStats {
+    uint64_t reads;
+    uint64_t writes;
+    uint64_t bank_stalls;
+
+    PerfStats()
+      : reads(0)
+      , writes(0)
+      , bank_stalls(0)
+    {}
+
+    PerfStats& operator+=(const PerfStats& rhs) {
+      this->reads += rhs.reads;
+      this->writes += rhs.writes;
+      this->bank_stalls += rhs.bank_stalls;
+      return *this;
+    }
+  };
+
+  std::vector<SimPort<MemReq>> Inputs;
+  std::vector<SimPort<MemRsp>> Outputs;
+
+  TensorMem(const SimContext& ctx, const char* name, const Config& config);
+  virtual ~TensorMem();
+
+  void reset();
+
+  void read(void* data, uint64_t addr, uint32_t size);
+
+  void write(const void* data, uint64_t addr, uint32_t size);
+
+  void tick();
+
+  const PerfStats& perf_stats() const;
+
+protected:
+
+  class Impl;
+  Impl* impl_;
+};
+
+}
diff --git a/sim/simx/types.h b/sim/simx/types.h
index 3182c33..cfe115b 100644
--- a/sim/simx/types.h
+++ b/sim/simx/types.h
@@ -708,7 +708,8 @@ using IntrArgs = std::variant<
 enum class AddrType {
   Global,
   Shared,
-  IO
+  IO,
+  Tensor
 };
 
 inline AddrType get_addr_type(uint64_t addr) {
@@ -720,6 +721,11 @@ inline AddrType get_addr_type(uint64_t addr) {
         return AddrType::Shared;
     }
   }
+  if (EXT_TCU_ENABLED) {
+    if (addr >= TMEM_BASE_ADDR && addr < TMEM_END_ADDR) {
+      return AddrType::Tensor;
+    }
+  }
   return AddrType::Global;
 }
 
@@ -728,6 +734,7 @@ inline std::ostream &operator<<(std::ostream &os, const AddrType& type) {
   case AddrType::Global: os << "Global"; break;
   case AddrType::Shared: os << "Shared"; break;
   case AddrType::IO:     os << "IO"; break;
+  case AddrType::Tensor: os << "Tensor"; break;
   default: assert(false);
   }
   return os;
@@ -1583,6 +1590,8 @@ public:
   SimPort<LsuReq> ReqDC;
   SimPort<LsuRsp> RspDC;
 
+  // TODO: Add ReqTmem, RspTmem ports for TCU local memory access
+
   LocalMemSwitch(
     const SimContext& ctx,
     const char* name,
